walkthroughs/howto-mutual-tls-file-provided-by-acm/README.md

```diff
@@ -0,0 +1,361 @@
+
+> Note: We do this first to illustrate what happens when TLS validation is met. 
+
+```bash
+./infrastructure/mesh.sh mtls
+```
+
+This updates the virtual gateway with a `certificate` object.
+
+```yaml
+meshName: mtls-appmesh
+virtualGatewayName: gateway-vgw
+spec:
+  backendDefaults:
+    clientPolicy:
+      tls:
+        certificate:
+          file:
+            certificateChain: /keys/colorgateway_endpoint_cert_chain.pem
+            # this is the ACM Certificate Chain for colorgateway.mtls.svc.cluster.local
+            privateKey: /keys/colorgateway_endpoint_dec_pri_key.pem
+            # this is the decrypted ACM private key for colorgateway.mtls.svc.cluster.local
+        enforce: true
+        validation:
+          trust:
+            acm:
+              certificateAuthorityArns:
+                - >-
+                  arn:aws:acm:us-east-1:012345678912:certificate/aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa
+              # this is the ACM ARN for colorteller.mtls.svc.cluster.local
+  listeners:
+    - portMapping:
+        port: 9080
+        protocol: http
+```
+
+On the virtual node `validation` is added to check for colorgateway certificate chain.
+```yaml
+meshName: mtls-appmesh
+virtualNodeName: colorteller-vn
+spec:
+  listeners:
+    - portMapping:
+        port: 9080
+        protocol: http
+      tls:
+        certificate:
+          acm:
+            certificateArn: >-
+              arn:aws:acm:us-east-1:012345678912:certificate/aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa
+            # this is the ACM ARN for colorteller.mtls.svc.cluster.local
+        mode: STRICT
+        validation:
+          trust:
+            file:
+              certificateChain: /keys/colorgateway_endpoint_cert_chain.pem
+              # this is the ACM Certificate Chain for colorgateway.mtls.svc.cluster.local
+  serviceDiscovery:
+    awsCloudMap:
+      namespaceName: mtls.svc.cluster.local
+      serviceName: colorteller
+```
+
+Recycle the containers after the update so that a new connection can take place. Then wait for the deployment to complete.
+```bash
+aws ecs update-service --cluster $ECS_CLUSTER --service $COLORTELLER_SERVICE --force-new-deployment
+aws ecs update-service --cluster $ECS_CLUSTER --service $GATEWAY_SERVICE --force-new-deployment
+aws ecs wait services-stable --cluster $ECS_CLUSTER --service $COLORTELLER_SERVICE
+aws ecs wait services-stable --cluster $ECS_CLUSTER --service $GATEWAY_SERVICE
+```
+
+Now query the endpoint again
+```bash
+% curl $COLORAPP_ENDPOINT
+yellow
+```
+
+From the bastion, take a look at SSL stats:
+
+```bash
+% curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
+listener.0.0.0.0_15000.ssl.handshake: 1
+listener.0.0.0.0_15000.ssl.no_certificate: 0
```

Comment:
So I'm getting incosistent behavior here. First time I run the command I get this output but the next time I get 9/8. Like below:

```
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 9
listener.0.0.0.0_15000.ssl.no_certificate: 8
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 1
listener.0.0.0.0_15000.ssl.no_certificate: 0
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 9
listener.0.0.0.0_15000.ssl.no_certificate: 8
sh-4.2$ curl -s colorteller.mtls.svc.cluster.local:9901/stats | grep -E 'ssl.handshake|ssl.no_certificate'
listener.0.0.0.0_15000.ssl.handshake: 1
listener.0.0.0.0_15000.ssl.no_certificate: 0
```

It seems to toggle back and forth. Can you explain this? / is this intended?

Reply:
I didn't wait for the deployment to complete. These requests were being load balanced over the tls and mtls deployments. Looks good!